### ISSUE TITLE
non-member false positive when subclass is used with with-statement

### ISSUE THREAD

**Issue by eight04** [PyCQA/pylint#3450]
### Steps to reproduce
```py
class A:
  def __enter__(self):
    return self
    
  def __exit__(self, *args):
    pass
    
class B(A):
  def foo(self):
    print("foo")

with B() as b:
  b.foo()

```
### Current behavior

```
test2.py:13:2: E1101: Instance of 'A' has no 'foo' member (no-member)
```

### Expected behavior

It works fine without the `with` statement:

```py
b = B()
b.foo()
```

### pylint --version output

pylint 2.4.4
astroid 2.3.3
Python 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 22:39:24) [MSC v.1916 32 bit (Intel)]

I guess the problem is that pylint thinks `return self` is always an instance of `A`?

**Comment by PCManticore** [comment]
This is no longer reproduceable with the latest unreleased versions, namely:
```
pylint 2.5.0-dev1
astroid 2.4.0
```

I remember fixing something similar a few weeks ago but can't find the exact commit, but if you have some time, please give it a go with `astroid==2.4.0` installed from the `astroid` master branch: https://github.com/PyCQA/astroid/
